packages/apollo-federation/src/composition/validate/sdl/uniqueTypeNamesWithoutEnumsOrScalars.ts

```diff
@@ -30,25 +50,128 @@ export function UniqueTypeNamesWithoutEnumsOrScalars(
+      if (diffLength > 0) {
+        context.reportError(
+          errorWithCode(
+            'VALUE_TYPE_UNION_TYPES_MISMATCH',
+            `The union '${typeName}' is defined in multiple places, however the unioned types do not match. Union types with the same name must also consist of identical types. The type${
+              diffLength > 1 ? 's' : ''
+            } ${unionDiff.map(diffEntry => diffEntry.name.value).join(', ')} ${
+              diffLength > 1 ? 'are' : 'is'
+            } mismatched.`,
+            [node, duplicateTypeNode],
+          ),
+        );
+      }
+
+      return false;
+    }
+
+    /*
+     * Return early for value types
+     * Value types:
+     * 1) have the same kind (type, interface, input), extensions are excluded
+     * 2) are not entities
+     * 3) have the same set of fields
+     */
+    if (duplicateTypeNode) {
+      const possibleErrors: GraphQLError[] = [];
+      // By inspecting the diff, we can warn when field types mismatch.
+      // A diff entry will exist when a field exists on one type and not the other, or if there is a type mismatch on the field
+      // i.e. { sku: [Int, String!], color: [String] }
+      const { kind, fields } = diffTypeNodes(node, duplicateTypeNode);
+
+      const fieldsDiff = Object.entries(fields);
+      const typesHaveSameShape =
+        fieldsDiff.length === 0 ||
+        fieldsDiff.every(([fieldName, types]) => {
+          // If a diff entry has two types, then the field name matches but the types do not.
+          // In this case, we can push a useful error to hint to the user that we
+          // think they tried to define a value type, but one of the fields has a type mismatch.
+          if (types.length === 2) {
+            possibleErrors.push(
+              errorWithCode(
+                'VALUE_TYPE_FIELD_TYPE_MISMATCH',
+                `Found field type mismatch on expected value type. '${typeName}.${fieldName}' is defined as both a ${types[0]} and a ${types[1]}. In order to define '${typeName}' in multiple places, the fields and their types must be identical.`,
+                [node, duplicateTypeNode],
+              ),
+            );
+            return true;
+          }
+          return false;
+        });
+
+      // Once we determined that types have the same shape (name, kind, and field
+      // names), we can provide useful errors
+      if (typesHaveSameShape) {
+        // Report errors that were collected while determining the matching shape of the types
+        possibleErrors.forEach(error => context.reportError(error));
+
+        // Error if the kinds don't match
+        if (kind.length > 0) {
+          context.reportError(
+            errorWithCode(
+              'VALUE_TYPE_KIND_MISMATCH',
+              `Found kind mismatch on expected value type. '${typeName}' is defined as both a ${kind[0]} and a ${kind[1]}. In order to define ${typeName} in multiple places, the kinds must be identical.`,
```

Comment:
Same here: can we use the logging util to print serviceName of one of the two services and say which service defined which kind. 

I'm also not sure people have a notion of what a `Kind` is. So maybe calling it a definition kind or something?

Reply:
Agree, it felt weird to refer to "kind" here, but I don't know if "definition kind" makes it more clear. Maybe by example?

> Found kind mismatch (type, interface, input) on expected value type...